internal/ceb/child_exec.go

```diff
@@ -14,23 +17,127 @@ import (
 // we run. This just sets the `childCmd` field on the CEB structure. This
 // does not have any side effecting behavior.
 func (ceb *CEB) initChildCmd(ctx context.Context, cfg *config) error {
+	// Prepare our base command. This validates that the path and everything is found.
 	cmd, err := ceb.buildCmd(ctx, cfg.ExecArgs)
 	if err != nil {
 		return err
 	}
+	ceb.childCmdBase = cmd
+
+	// Setup our channels and goroutine to prepare to execute this. This
+	// won't actually start any child commands, but it starts the watcher
+	// goroutine that will eventually run them.
+	childCmdCh := make(chan *exec.Cmd, 5)
+	doneCh := make(chan error, 1)
+	go ceb.watchChildCmd(ctx, childCmdCh, doneCh)
+	ceb.childCmdCh = childCmdCh
+	ceb.childDoneCh = doneCh
 
-	ceb.childCmd = cmd
 	return nil
 }
 
-// execChildCmd starts the child process, and waits for completion by sending an
+// watchChildCmd should be started in a goroutine. This will run the child
+// command sent on the channel. This ensures only one child command is run
+// at a time.
+func (ceb *CEB) watchChildCmd(
+	ctx context.Context,
+	cmdCh <-chan *exec.Cmd,
+	doneCh chan<- error,
+) {
+	// We always close the done channel when we exit so that callers can
+	// detect this and also exit.
+	defer close(doneCh)
+
+	log := ceb.logger.Named("watchChildCmd")
+	log.Debug("starting child command watch loop")
+
+	var currentCh <-chan error
+	var currentCmd *exec.Cmd
+	for {
+		select {
+		case <-ctx.Done():
+			log.Warn("request to exit")
+
+			// If we have a child, we need to exit that first.
+			if currentCh != nil {
+				log.Info("terminating current child process")
+				err := ceb.termChildCmd(log, currentCmd, currentCh, false)
+				log.Info("child process termination result", "err", err)
+			}
+
+			return
+
+		case cmd := <-cmdCh:
+			log.Info("child command received")
+
+			if currentCh == nil {
+				// First client is always the default, so let's wait for
```

Comment:
What does 'the default' here mean? This whole comment feels confusing, I guess we're waiting for 2 seconds or if we got a different command to run?

Reply:
Yeah I’ll expand on this comment. I might do this in a more robust way too. But basically we always send the “default” command which is just the basic child process as invoked on the command line. But we expect to get another one quickly if we connect to the server with env vars and such. 

I might just do this with a more explicitly triggered system though so it’s not just timer crossing fingers

Reply:
Made this better, triggered.

Reply:
If we're going to start the command and then kill it a second later, is it worth having the system wait until config arrives before starting then? 150 milliseconds isn't very long.

Reply:
Let me just switch this out for a non-timer. I think this is giving the wrong idea here. This is really just draining the channel buffer.

Reply:
Done! Take another look. Should be more clear.

Reply:
Ahhhhhh, I see now!